csrc/multidevice/utils.cpp

```diff
@@ -267,17 +274,33 @@ void insertReshardings(Fusion* fusion) {
         "multi-output expressions are not supported");
     auto output = expr->outputs().at(0)->as<TensorView>();
     std::vector<TensorView*> new_inputs;
-    for (auto input : getTvsWithDifferentSharding(
-             output, ir_utils::filterByType<TensorView>(expr->inputs()))) {
-      // TODO: reuse cacheAfter?
-      // TODO: here we should add a mechanism to potentially reuse the inserted
-      // resharding accross all the consumer of the resharded tensor. This way
-      // we could avoid wasteful resharding set insertion.
-      TensorView* new_input = set(input);
-      new_inputs.push_back(new_input);
-      expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
+    auto inputs = getTvsWithDifferentSharding(
+        output, ir_utils::filterByType<TensorView>(expr->inputs()));
+
+    // Insert resharding set after the expr when there is only one input.
+    // input [expr] output [set] new_output
+    if (!inputs.empty() && expr->inputs().size() == 1) {
```

Comment:
Does it make sense to unit-test the change in insertReshardings, e.g., in tests/cpp/test_multidevice_sharding.cpp? These tests should run just fine on a single-device system so regressions are easier to discover.

Reply:
Yeah I should probably test this... I think I'll pull this change out into a separate PR with some tests.

Reply:
Moved to a separate PR. This one is updated with explicit sets so that it doesn't need the changes in `insertReshardings`